core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -478,6 +478,45 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
   protected def minRegisteredRatio: Double = _minRegisteredRatio
 
+  @GuardedBy("CoarseGrainedSchedulerBackend.this")
+  private val executorsToDecommissionInBatches = new HashSet[String]
+
+  private val executorDecommissionEnabled = conf.get(config.DECOMMISSION_ENABLED)
+  private var periodicBatchExecutorsDecommissionThread: Option[ScheduledExecutorService] = None
+  private var periodicBatchExecutorsDecommissionTask: Option[ScheduledFuture[_]] = None
+
+  if (executorDecommissionEnabled) {
+    /**
+     * Start a thread which runs a task periodically to decommission executors.
```

Comment:
There's only a limited time (e.g., 30s in Azure) to decommission executors. Otherwise, executors can shut down before migrating data.

Reply:
I have kept batch decommission interval as 3s and batch size to Int.MaxValue for now. We won't regress with default settings. 
However we can set these variables aggressively to try to migrate data from all decommissioning executors.

Reply:
I think it could make sense for us to distinguish between executors where Spark has chosen decommissioning and we are not under a time pressure and those where the cloud provider / cluster manager has trigger decommissioning and we are under a time pressure. In the second case I don't think we should apply any rate limiting.

Reply:
... and prioritization if possible. If Spark has chosen to decommission but in the mean time the cluster forces several less optimal nodes to die, it may produce more work that no longer requires the Spark-chosen executor to decommission any more.